src/testing/slim_check/gen.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Simon Hudon
+-/
+
+import control.uliftable
+import system.random
+import system.random.basic
+
+/-!
+# Gen Monad
+
+This monad is used to formulate randomized computations with a parameter
+to specify the desired size of the result.
+
+This is a port of the Haskell QuickCheck library.
+
+## Main definitions
+  * `gen` monad
+
+## Local notation
+
+ * `i .. j` : `Icc i j`, the set of values between `i` and `j` inclusively;
+
+## Tags
+
+random testing
+
+## References
+
+  * https://hackage.haskell.org/package/QuickCheck
+
+-/
+
+universes u v
+
+namespace slim_check
+
+/-- Monad to generate random examples to test properties with.
+It has a `nat` parameter so that the caller can decide on the
+size of the examples -/
+@[reducible, derive [monad, is_lawful_monad]]
+def gen (α : Type u) := reader_t (ulift ℕ) rand α
+
+variable (α : Type u)
+
+local infix ` .. `:41 := set.Icc
+
+/-- execute a `gen` inside the `io` monad using `i` as the example
+size and creating a fresh random number generator -/
+def io.run_gen {α} (x : gen α) (i : ℕ) : io α :=
+io.run_rand (x.run ⟨i⟩)
+
+namespace gen
+
+section rand
+
+variables [preorder α]
+
+/-- lift `random.random` to the `gen` monad -/
+def choose_any [random α] : gen α :=
+⟨ λ _, rand.random α ⟩
+
+variables {α}
+
+/-- lift `random.random_r` to the `gen` monad -/
+def choose [bounded_random α] (x y : α) (p : x ≤ y) : gen (x .. y) :=
+⟨ λ _, rand.random_r x y p ⟩
+
+end rand
+
+open nat (hiding choose)
+
+/-- generate a `nat` example between `x` and `y` -/
+def choose_nat (x y : ℕ) (p : x ≤ y) : gen (x .. y) :=
+choose x y p
+
+open nat
+
+instance : uliftable gen.{u} gen.{v} :=
+reader_t.uliftable' (equiv.ulift.trans equiv.ulift.symm)
+
+instance : has_orelse gen.{u} :=
+⟨ λ α x y, do
+  b ← uliftable.up $ choose_any bool,
+  if b.down then x else y ⟩
+
+variable {α}
+
+/-- Get access to the size parameter of the `gen` monad. For
+reasons of universe polymorphism, it is specified in
+continuation passing style. -/
+def sized (cmd : ℕ → gen α) : gen α :=
+⟨ λ ⟨sz⟩, reader_t.run (cmd sz) ⟨sz⟩ ⟩
+
+/-- create `n` examples using `cmd` -/
+def vector_of : ∀ (n : ℕ) (cmd : gen α), gen (vector α n)
+| 0 _ := return vector.nil
+| (succ n) cmd := vector.cons <$> cmd <*> vector_of n cmd
+
+/-- create a list of examples using `cmd`. The size is controlled
+by the size parameter of `gen` -/
+def list_of (cmd : gen α) : gen (list α) :=
+sized $ λ sz, do
+do ⟨ n ⟩ ← uliftable.up $ choose_nat 0 (sz + 1) dec_trivial,
+   v ← vector_of n.val cmd,
+   return v.to_list
+
+open ulift
+
+/-- given a list of example generators, choose one to create an example -/
+def one_of (xs : list (gen α)) (pos : fact $ 0 < xs.length) : gen α := do
```

Comment:
```suggestion
def one_of (xs : list (gen α)) (pos : fact $ 0 < xs.length) : gen α := do
```

Having `fact` as an explicit argument seems backwards.  The `fact` gadget is there to wrap propositions as type classes.  Either make an explicit argument without `fact`, or make the `fact` instance-implicit.

Reply:
Without `fact`, the code becomes:

```lean
def one_of (xs : list (gen α)) (pos : 0 < xs.length) : gen α := do
have fact $ 0 < xs.length, from pos,
n ← uliftable.up $ choose_any (fin xs.length),
list.nth_le xs (down n).val (down n).is_lt
```

I need the `fact` gadget so that `choose_any` will find an instance `random (fin xs.length)`. I thought we could dispense with the superfluous `have`. And I don't think `[fact $ 0 < xs.length]` is a good idea because, as you used it above, using `one_of` with a literal list which we'd be forced to reproduce in a `have` close in the calling code.

Reply:
I think the explicit version without `fact` is still better than the current one, even if it's just for consistency.  We don't want to put `fact $` in front of every proposition in mathlib.